File: coding_int/chapter_2/2_6_Palindrome.py
class node:
    def __init__(self, value=None):
        self.next = None
        self.datavalue = value
    
class llist:
    def __init__(self, head=None):
        self.head = head
        self.tail = None

    def __iter__(self):
        current = self.head
        while(current):
            yield current
            current=current.next

    def __str__(self):
        current = self.head
        result = ""
        while(current):
            result = result + current.datavalue + " " 
            current = current.next

        return result

    def addValue(self, dataitem):

        new_node = node(dataitem)
        if self.head is None:
            self.head = new_node
            self.tail = new_node

        else:
            self.tail.next = new_node
            self.tail = new_node

    def isPalindrome(self):

        forward = ''.join([x.datavalue for x in self])
        backward = ''.join([x.datavalue for x in self])[::-1]

        return(forward == backward)
        

if __name__ == "__main__":

    test_list = llist()
    test_list.addValue("c")
    test_list.addValue("a")
    test_list.addValue("a")
    test_list.addValue("c")
    print(test_list)
    print(test_list.isPalindrome())

    test_list_2 = llist()
    test_list_2.addValue("c")
    test_list_2.addValue("a")
    test_list_2.addValue("b")
    test_list_2.addValue("c")
    test_list_2.addValue("a")
    print(test_list_2)
    print(test_list_2.isPalindrome())

            

File: coding_int/chapter_2/2_6_Palindrome.cpp
#include<iostream>
#include<string>

class node {
    public:

    node* next;
    std::string data;
};

class llist {

    private:

    node* head;
    node* tail;

    public:

    llist() {
        head = NULL;
        tail = NULL;
    }

    void addValue(std::string dataitem) {

        node *temp = new node;
        temp->data = dataitem;

        if (head == NULL) {
            head=temp;
            tail=temp;
        } else {
            tail->next = temp;
            tail = temp;
        }
    }

    void printItems() {

        node *temp = new node;
        temp = head;

        while(temp) {
            std::cout << temp->data << " ";
            temp = temp->next;
        }

        std::cout << std::endl;
    }

    void isPalindrome(){

        std::string forward = "";
        std::string backward = "";

        node *temp = new node;
        temp = head;

        while (temp) {
            forward = forward + temp->data;
            backward = temp->data + backward;
            temp = temp->next;
        }

        bool result = (forward==backward);
        std::cout << "is palindrome? " << result << std::endl;
    }
};

int main() {

    llist *test_1 = new llist();
    test_1->addValue("c");
    test_1->addValue("a");
    test_1->addValue("a");
    test_1->addValue("c");
    test_1->printItems();
    test_1->isPalindrome();

    llist *test_2 = new llist();
    test_2->addValue("c");
    test_2->addValue("a");
    test_2->addValue("b");
    test_2->addValue("c");
    test_2->addValue("a");
    test_2->printItems();
    test_2->isPalindrome();

}